lib/src/lock/unix.rs

```diff
@@ -45,17 +45,29 @@ impl FileLock {
                 }
             })?;
 
-            let stat = rustix::fs::fstat(&file).map_err(|errno| FileLockError {
-                message: "failed to stat lock file",
-                path: path.clone(),
-                err: errno.into(),
-            })?;
-            if stat.st_nlink == 0 {
-                // Lockfile was deleted, probably by the previous holder's `Drop` impl; create a
-                // new one so our ownership is visible, rather than hidden in an
-                // unlinked file. Not always necessary, since the previous
-                // holder might have exited abruptly.
-                continue;
+            match rustix::fs::fstat(&file) {
+                Ok(stat) => {
+                    if stat.st_nlink == 0 {
+                        // Lockfile was deleted, probably by the previous holder's `Drop` impl;
+                        // create a new one so our ownership is visible,
+                        // rather than hidden in an unlinked file. Not
+                        // always necessary, since the previous holder might
+                        // have exited abruptly.
+                        continue;
+                    }
+                }
+                Err(rustix::io::Errno::STALE) => {
+                    // The file handle is stale. This can happen on NFS filesystems
+                    // and should be treated as if the lockfile was deleted.
```

Comment:
Maybe note that this probably indicates the file was deleted and recreated remotely?